### DISCUSSION TITLE
Multivariate Time Series Transformer

### DISCUSSION THREAD

**Discussion by sergiofdez02** [community/community#130066] [Programming Help]
### Body

class Transformer(tf.keras.Model):
    def __init__(self, num_blocks_enc, num_blocks_dec, d_model, d_ff, num_heads, output_size, input_width, label_width, rate=0.1, **kwargs):
        super(Transformer, self).__init__(**kwargs)
        self.time2vec = Time2Vec(d_model)
        self.encoder = [EncoderLayer(d_model, d_ff, num_heads, rate) for _ in range(num_blocks_enc)]
        self.decoder = [DecoderLayer(d_model, d_ff, num_heads, rate) for _ in range(num_blocks_dec)]
        self.final_layer = tf.keras.layers.Dense(output_size, activation='relu')
        self.num_blocks_enc = num_blocks_enc
        self.num_blocks_dec = num_blocks_dec
        self.d_model = d_model
        self.d_ff = d_ff
        self.num_heads = num_heads
        self.output_size = output_size
        self.rate = rate
        self.input_width = input_width
        self.label_width = label_width  # set label_width as an instance variable
    
    def call(self, inputs, training=True):
        if isinstance(inputs, tuple):
            inputs, labels = inputs
        else:
            labels = None

        enc_padding_mask = create_padding_mask(inputs)
        x = self.time2vec(inputs)
        
        for i in range(len(self.encoder)):
            x = self.encoder[i](x, training=training, mask=enc_padding_mask)
        enc_output = x

        if labels is not None:
            # Training phase
            dec_padding_mask = create_padding_mask(labels)
            look_ahead_mask = create_look_ahead_mask(tf.shape(labels)[1])
            combined_mask = tf.maximum(dec_padding_mask, look_ahead_mask)
            x = self.time2vec(labels)
            dec_out = x
            
            for i in range(len(self.decoder)):   
                x = self.decoder[i](x, enc_output, training=training, look_ahead_mask=combined_mask, padding_mask=enc_padding_mask)
        else:
            # Inference phase
            x = inputs[:, -1:, :]  # Use the last few inputs as the initial decoder input
            dec_out = tf.zeros((tf.shape(x)[0], 1, self.d_model))  # Initialize decoder output with the same last dimension as the decoder output

            x = self.time2vec(x)
            
            for _ in range(self.label_width-1):  # Generate output sequence of the same length as labels
                dec_padding_mask = create_padding_mask(x)
                look_ahead_mask = create_look_ahead_mask(tf.shape(x)[1])
                combined_mask = tf.maximum(dec_padding_mask, look_ahead_mask)
                
                
                for i in range(len(self.decoder)):   
                    x = self.decoder[i](x, enc_output, training=training, look_ahead_mask=combined_mask, padding_mask=enc_padding_mask)
                dec_out = tf.concat([dec_out, x], axis=1)  # Append output to decoder output sequence
                
        final_output = self.final_layer(dec_out)
        
        return final_output
        
 Is this implementation correct?
